harness/determined/_training.py

```diff
@@ -0,0 +1,113 @@
+        latest_total_batches = latest.get("totalBatches")
+        log.debug(f"get_last_validation() -> {latest_total_batches}")
+        return latest_total_batches
+
+    def report_training_metrics(
+        self,
+        total_batches: int,
+        metrics: Dict[str, Any],
+        *,
+        batch_metrics: Optional[List] = None,
+        total_records: Optional[int] = None,
+        total_epochs: Optional[int] = None,
+        computed_records: Optional[int] = None,
+    ) -> None:
+        body = {
+            "trial_run_id": self._run_id,
+            "total_batches": total_batches,
+            "metrics": metrics,
+        }
+        if batch_metrics is not None:
+            body["batch_metrics"] = batch_metrics
+        if total_records is not None:
+            body["total_records"] = total_records
+        if total_epochs is not None:
+            body["total_epochs"] = total_epochs
+        if computed_records is not None:
+            body["computed_records"] = computed_records
+        log.info(f"report_training_metrics(total_batches={total_batches}, metrics={metrics})")
+        self._session.post(
+            f"/api/v1/trials/{self._trial_id}/training_metrics",
+            encoded_body=det.util.json_encode(body),
+        )
+
+    def report_validation_metrics(
+        self,
+        total_batches: int,
+        metrics: Dict[str, Any],
+        *,
+        total_records: Optional[int] = None,
+        total_epochs: Optional[int] = None,
+        computed_records: Optional[int] = None,
+    ) -> None:
+        body = {
+            "trial_run_id": self._run_id,
+            "total_batches": total_batches,
+            "metrics": metrics,
+        }
+        if total_records is not None:
+            body["total_records"] = total_records
+        if total_epochs is not None:
+            body["total_epochs"] = total_epochs
+        if computed_records is not None:
+            body["computed_records"] = computed_records
+        log.info(f"report_validation_metrics(total_batches={total_batches}, metrics={metrics})")
+        self._session.post(
+            f"/api/v1/trials/{self._trial_id}/validation_metrics",
+            encoded_body=det.util.json_encode(body),
+        )
+
+    def report_early_exit(self, reason: EarlyExitReason) -> None:
+        body = {"reason": EarlyExitReason(reason).value}
+        log.info(f"report_early_exit({reason})")
+        self._session.post(
+            f"/api/v1/trials/{self._trial_id}/early_exit",
+            encoded_body=det.util.json_encode(body),
+        )
+
+    def get_experiment_best_validation(self) -> Optional[float]:
+        log.debug("get_experiment_best_validation()")
+        try:
+            r = self._session.get(
+                f"/api/v1/experiments/{self._exp_id}/searcher/best_searcher_validation_metric"
+            )
+        except errors.NotFoundException:
+            # 404 means 'no validations yet'.
```

Comment:
Talk to @stoksc.  I voted for an in-json response.  He disagreed.  We decided that it wasn't an external API so we didn't need to fight over it until we were ready to version the master-harness REST api.

Reply:
wait, or maybe that disagreement was about the searcher metrics.  I can't remember anymore.  But in any case we can change it if we all agree on what it should be, or we can punt until later if we have something that works but we don't agree yet.

Reply:
It was a different API but I mean.. it works.. `/api/v1/experiments/{self._exp_id}/searcher/best_searcher_validation_metric -> 404` means the best_searcher_validation_metric resource for exp_id = something is not found. But I dont care a lot, it's just returning None from grpc is needlessly weird (blame on grpc there).. we could use an optional, not sure what grpc -> grpc_gateway ends up encoding those as.

Reply:
Is this a case where we should use echo instead of grpc?  Where have we landed on that discussion?

I would much rather pick a REST API we like and use echo if necessary than choose a REST API we don't really like in order to accommodate a tool that we presently have the option to not use.